### PR TITLE
Make improvements to `retrieve-codeowners` tool: change accepted params (`targetPath` etc.) + refactor

### PR DESCRIPTION
**Pull request by konrad-jamrozik**
This PR is part of work required in preparation to merge:
- #5088

Specifically, to enable review of ownership changes due to upcoming wildcards support, as explained in:
- https://github.com/Azure/azure-sdk-tools/pull/5088#issuecomment-1376542202

As such, this PR contributes to:
- #2770

Following PR depends on this one:
- #5103

## Changes made

All changes are within the scope of following paths:
- `/tools/code-owners-parser/**`
- `/tools/identity-resolutions/Services/GitHubService.cs`
- `/tools/notification-configurator/**`

Behavior changes:
- `RetrieveCodeOwners.Program` now accepts:
    - `targetPath` instead of `targetDirectory` 
    - `excludeNonUserAliases` instead of `filterOutNonUserAliases`
    - `codeownersFilePathOrUrl` instead of `codeownersFilePath`
    - the internal implementation didn't change for any of the params; just the param names were misleading
- Added `Debug.Assert(targetPath != null);` precondition to `CodeownersFile.GetMatchingCodeownersEntry`
- Parsing of CODEOWNERS line now will throw `InvalidOperationException` on case that should be impossible.

Refactoring changes:
- Comments: documented `RetrieveCodeOwners.Program`.
- Renames: made names consistent by changing all `[c|C]odeOwner` occurrences (except namespaces) to `[c|C]odeowners`. E.g. `CodeOwnerEntry` to `CodeownersEntry`.
- Renamed `MainTests` to `ProgramTests` to follow the test naming convention of `<class>Test` not `<method>Tests`.
- Addressed various IDE code quality rules (`CA*`) warnings: https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/
- Addressed various IDE code style rules (`IDE*`) warnings: https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/
- Made method names more consistent and predictable	in `CodeownersFile`.
- Extracted some intention-revealing methods, like `ProcessCodeownersLine` or `IsCommentLine`.
- Reordered relevant params so that `targetPath` always comes first, before `codeownersEntries`.
- Improved method and param names in `Azure.Sdk.Tools.RetrieveCodeOwners.Tests.ProgramTests`.
- Broke lines at 120 chars.
- Various other minor renames, method extractions and refactorings.

### CODE REVIEW

File: tools/code-owners-parser/CodeOwnersParser/CodeownersFile.cs [hunk 1]
```diff
@@ -0,0 +1,115 @@
+using System;
```
Comment (konrad-jamrozik) [inline tools/code-owners-parser/CodeOwnersParser/CodeownersFile.cs:1]: FYI: this file was renamed from `CodeOwnersFile` but git didn't pick up on that.

File: tools/code-owners-parser/Azure.Sdk.Tools.RetrieveCodeOwners.Tests/ProgramTests.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
+using System.Collections.Generic;
```
Comment (konrad-jamrozik) [inline tools/code-owners-parser/Azure.Sdk.Tools.RetrieveCodeOwners.Tests/ProgramTests.cs:1]: FYI this is renamed from `MainTests`.

File: tools/code-owners-parser/Azure.Sdk.Tools.RetrieveCodeOwners/Program.cs [hunk 3]
```diff
@@ -5,36 +5,47 @@
...
+        /// "new CodeownersEntry()" if no path in the CODEOWNERS data matches.
+        /// <br/><br/>
+        /// From the Main method: exit code. 0 if successful, 1 if error.
+        /// </returns>
         public static int Main(
-            string codeOwnerFilePath,
-            string targetDirectory,
-            bool filterOutNonUserAliases = false
-            )
+            string targetPath,
```
Comment (weshaggard) [inline tools/code-owners-parser/Azure.Sdk.Tools.RetrieveCodeOwners/Program.cs:27]: Just keep in mind that you will need to update places like https://github.com/Azure/azure-sdk-tools/blob/main/eng/common/scripts/get-codeowners.ps1 when you update to the latest package version.

**Review by weshaggard — APPROVED** [review]

File: tools/code-owners-parser/Azure.Sdk.Tools.RetrieveCodeOwners/Program.cs (hunk 3 above)
Comment (konrad-jamrozik) [reply tools/code-owners-parser/Azure.Sdk.Tools.RetrieveCodeOwners/Program.cs:27]: @weshaggard yeah I already have a PR for that (in draft state as of now), thanks for the reminder! :)
- #5103

**Pull request merged**